# Table of Contents

01. Importing libraries
02. Importing data
03. Consistency checks on df_prods dataframe
    a) Descriptive stats
    b) Missing values check
    c) Duplicate values check
04. Consistency checks on df_ords dataframe
    a) Descriptive stats
    b) Mixed-type data check
    c) Missing values check
    d) Duplicate values check
05. Exporting dataframes

## 01. Importing libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

## 02. Importing data

In [2]:
path=r'/Users/tejas/Documents/CareerFoundry_DataAnalytics/Data Immersion/A4/Instacart Basket Analysis'

In [3]:
# import products.csv data set
df_prods=pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'), index_col=False)

In [4]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [5]:
# import orders_wrangled.csv data set
df_ords=pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'), index_col=False)

In [6]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

## 03. Consistency checks on df_prods dataframe

### a) Descriptive stats

In [67]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### The maximum price of 99999 seems odd for a grocery store item.

### b) Missing values check

In [7]:
# check for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [8]:
# create new df consisting of only missing values
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [9]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [10]:
# check number of rows in df_prods
df_prods.shape

(49693, 5)

In [11]:
# create new df consisting of non-missing values
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [12]:
# check number of rows in df_prods_clean
df_prods_clean.shape

(49677, 5)

### c) Duplicate values check

In [13]:
# create a new df consisting of only duplicate values
df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [14]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [15]:
# check number of rows in df_prods_clean
df_prods_clean.shape

(49677, 5)

In [16]:
# create new df consisting of unique values by dropping duplicates
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [17]:
# check number of rows in df_prods_clean_no_dups
df_prods_clean_no_dups.shape

(49672, 5)

## 04. Consistency checks on df_ords dataframe

### a) Descriptive stats

In [60]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

#### The above stats look fine to me.

### b) Mixed-type data check

In [21]:
# check for mixed type values
for col in df_ords.columns.tolist():
    weird=(df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_ords[weird])>0:
        print(col)

### c) Missing values check

In [22]:
# check for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### days_since_prior_order column has 206209 missing values. Any order placed for the first time will have a missing value in the days_since_prior_order column.
#### PFB steps of how I plan to address this issue:
#### 1. Checking to make sure the frequency of missing values is same as the frequency of order_number=1.
#### 2. Create a new column (first_order) to flag the reason for missing values.

In [30]:
# create a new df consisting of only missing values
df_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [31]:
df_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [32]:
# check the frequency of order_number=1 in df_nan
df_nan['order_number'].value_counts()

1    206209
Name: order_number, dtype: int64

In [49]:
# check the frequency of order_number=1 in df_ords
df_ords['order_number'].value_counts()

1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: order_number, Length: 100, dtype: int64

In [50]:
# create a new function that checks if it is the first order or no
def is_first_order (row):
    if row['order_number']==1:
        return True
    else:
        return False

In [51]:
# apply the is_first_order function
df_ords.apply (lambda row: is_first_order(row), axis=1)

0           True
1          False
2          False
3          False
4          False
           ...  
3421078    False
3421079    False
3421080    False
3421081    False
3421082    False
Length: 3421083, dtype: bool

In [52]:
# create a new column in df_ords to flag the reason for missing values in days_since_prior_order
df_ords['first_order']=df_ords.apply(lambda row: is_first_order(row), axis=1)

In [53]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  first_order  
0                  8                     NaN         True  
1                  7                    15.0        False  
2                 12                    21.0        False  
3                  7                    29.0        False  
4                 15                    28.0        False

In [54]:
# check the frequency of first_order=True
df_ords['first_order'].value_counts()

False    3214874
True      206209
Name: first_order, dtype: int64

### d) Duplicate values check

In [55]:
# create a new df for duplicate values
df_ords_dups=df_ords[df_ords.duplicated()]

In [56]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

In [59]:
df_ords_dups.shape

(0, 8)

#### There are no duplicate values in df_ords

## 05. Exporting dataframes

In [ ]:
# export df_prods_clean_no_dups
df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data','Prepared Data','products_checked.csv'))

In [66]:
# export df_ords
df_ords.to_csv(os.path.join(path,'02 Data','Prepared Data','orders_checked.csv'))